packages/@aws-cdk/aws-s3/README.md

```diff
@@ -616,3 +616,41 @@ const bucket = new s3.Bucket(this, 'MyBucket', {
   }]
 });
 ```
+
+## Object Lock Configuration
+
+[Object Lock](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock-overview.html)
+can be configured to enable a write-once-read-many model for an S3 bucket. Object Lock must be
+configured when a bucket is created; if a bucket is created without Object Lock, it cannot be
+enabled later via the CDK.
+
+Object Lock can be enabled on an S3 bucket by specifying:
+
+```ts
+const bucket = new s3.Bucket(this, 'MyBucket', {
+  objectLock: {
+    enabled: true
+  }
+});
+```
+
+Usually, it is desired to not just enable Object Lock for a bucket but to also configure a
+[retention mode](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock-overview.html#object-lock-retention-modes)
+and a [retention period](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock-overview.html#object-lock-retention-periods).
+These can be specified by providing a `defaultRetention`:
+
+```ts
+const bucket = new s3.Bucket(this, 'MyBucket', {
+  objectLock: {
+    enabled: true,
+    defaultRetention: {
+      mode: s3.ObjectLockMode.GOVERNANCE,
+      duration: cdk.Duration.days(7 * 365),
+    }
+  }
+});
+
+```
+
+Note that if a `duration` is specified, it must be at least one day. Additionally, it is not
+permitted to specify a retention configuration without also enabling Object Lock.
```

Comment:
You state:

> It is possible to enable Object Lock without specifying a default duration.

These docs: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-objectlockconfiguration.html#cfn-s3-bucket-objectlockconfiguration-objectlockenabled state

> If Object Lock is turned on, bucket settings require both Mode and a period of either Days or Years.

I'm not sure which is true (it would be nice if you could confirm via a CFN deployment), but since object locks are targeting users with strict compliance requirements, I'm figuring that the most common use case will be to have the lock enabled with the duration and mode specified. 

With that in mind, can we remove the `objectLock` boolean property entirely? Since CFN refuses to allow you to set `ObjectLockEnabled` without the `Duration` and `Mode`, having `ObjectLockEnabled: false` prevents you from setting the `Duration` and `Mode`, we can set `ObjectLockEnabled` to `true` if `Duration` and `Mode` are specified and `false` otherwise.